src/types/module.rs

```diff
@@ -348,7 +348,7 @@ impl PyModule {
     ///
     /// [1]: crate::prelude::pyfunction
     /// [2]: crate::wrap_pyfunction
-    pub fn add_function<'a>(&'a self, fun: &'a PyCFunction) -> PyResult<()> {
+    pub fn add_function<'a>(&'a self, fun: Bound<'a, PyCFunction>) -> PyResult<()> {
```

Comment:
Maybe here we could just take `fun: impl ToPyObject`? That would avoid breaking existing direct calls, and we plan to remove this function anyway.

We could add a note to document the strange generic signature is for backwards compatibility while this API is on its way out.

Might make the implementation a little messier, but I'm sure it could work.

Reply:
Either way, by changing this function I think we need a `3808.changed.md` newsfragment.

Reply:
The main problem is what do to with `wrap_pyfunction!`, and whether we want to change it.

At the moment we have:

```rust
#[pymodule]
fn my_module(py: Python, m: &PyModule) {
    m.add_function(wrap_pyfunction!(foo, m)?)?;
}
```

If we don't change `wrap_pyfunction`, and go for `wrap_pyfunction_bound`, then we end up with this:

```rust 
#[pymodule]
fn my_module(py: Python, m: &Bound<PyModule>) {
    m.add_function(wrap_pyfunction_bound!(foo, m)?)?;
}
```

Which creates a bigger diff the more functions the user has. I guess it's not _too_ bad, though.

If we change `wrap_pyfunction!` return type like like we changed `intern!`, then the only thing users would need to change is the module type:

```rust 
#[pymodule]
fn my_module(py: Python, m: &Bound<PyModule>) {
    m.add_function(wrap_pyfunction!(foo, m)?)?;
}
```

... but maybe we're making this more complicated than it needs to be, and we should just add `wrap_pyfunction_bound`? I guess _unlike_ `intern!`, most `wrap_pyfunction!` calls will all be located in one place in the `#[pymodule]`. So it's not a very hard transformation.

... @snuderl what do you think of this? I guess adding `wrap_pyfunction_bound!` creates a bigger diff, but maybe it's not too bad?

Reply:
Yeah adding `wrap_pyfunction_bound!` should be pretty trivial it just hopped it could be avoided :)

There are 174 usages in the PyO3 repo, I guess we want to switch all of them right?

Reply:
Thinking about it, _probably_ adding `wrap_pyfunction_bound` gets very messy without the second commit I have in #3744, which allowed `&Bound<PyModule>` in `#[pymodule]`.  Because I think we would want to have have `wrap_pyfunction_bound!` used in `#[pymodule]` which use `&Bound<PyModule>`, and then maybe we keep just one test which uses the deprecated `&PyModule` to 

I really need to tidy that PR up and get it ready for review... 😢